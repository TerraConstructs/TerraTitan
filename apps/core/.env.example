GOOGLE_GENERATIVE_AI_API_KEY=
UPSTASH_TOKEN_SMALL=
UPSTASH_TOKEN_LARGE=
# # run local copy of signoz - see ../../signoz/README.md
# OTEL_EXPORTER_OTLP_ENDPOINT=http://localhost:4318
