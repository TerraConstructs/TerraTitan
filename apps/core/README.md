# Core

TerraTitan Core Workflows using Mastra.

[mastra](https://mastra.ai) is an open-source TypeScript agent framework

## Prerequisites

1. **Set up the Gemini API:**

   - Follow the Gemini API [Quickstart](https://ai.google.dev/gemini-api/docs/quickstart?lang=node)
   - Get API Key from [aistudio - apikey](https://aistudio.google.com/apikey)

1. **Set the Gemini API Key:**

   Set the `GOOGLE_GENERATIVE_AI_API_KEY` environment variable in the project: Create a `.env` file at the root of the project and add the following line (see `.env.exmaple` for reference):

   ```bash
   GOOGLE_GENERATIVE_AI_API_KEY=<your_api_key>
   ```

1. **To install dependencies:**

   ```bash
   pnpm install
   ```

1. **To run Mastra dashboard:**

   ```bash
   pnpm dev
   ```

1. **To Programmatically start the workflows defined under `src/index.ts`:**

   ```bash
   pnpm start
   ```

This project follow [mastra project structure](https://mastra.ai/docs/getting-started/project-structure)
