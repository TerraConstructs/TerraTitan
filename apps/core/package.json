{
  "name": "@terratitan/core",
  "version": "0.1.0",
  "type": "module",
  "private": true,
  "description": "LLM Workflow to convert AWS CDK Constructs to Terraform-CDK",
  "scripts": {
    "test": "vitest run",
    "check-types": "tsc --noEmit",
    "start": "tsx --no-deprecation --import ./src/instrumentation.ts src/index.ts",
    "lint": "eslint",
    "dev": "mastra dev",
    "build": "mastra build"
  },
  "keywords": [
    "terraform",
    "aws-cdk",
    "cdktf",
    "llm",
    "infrastructure-as-code"
  ],
  "author": "Charles Martinot and Vincent De Smet",
  "license": "GPL-3.0",
  "engines": {
    "node": ">=22.0.0"
  },
  "dependencies": {
    "@ai-sdk/cohere": "^1.2.7",
    "@ai-sdk/google": "^1.2.3",
    "@ai-sdk/openai": "^1.3.4",
    "@ai-sdk/provider": "^1.1.3",
    "@inquirer/prompts": "^7.4.1",
    "@mastra/core": "alpha",
    "@mastra/libsql": "alpha",
    "@mastra/rag": "alpha",
    "@mastra/upstash": "alpha",
    "@opentelemetry/exporter-logs-otlp-http": "0.57.2",
    "@opentelemetry/sdk-logs": "0.57.2",
    "adm-zip": "^0.5.16",
    "ai": "^4.3.9",
    "async-sema": "^3.1.1",
    "dotenv": "^16.4.7",
    "inquire-grouped-checkbox": "workspace:*",
    "jsii-reflect": "^1.111.0",
    "mastra-otel-logger": "workspace:*",
    "simple-git": "^3.27.0",
    "tiktoken": "^1.0.20",
    "ts-morph": "^16.0.0",
    "zod": "^3.24.2",
    "zodcli": "^0.0.4"
  },
  "peerDependencies": {
    "typescript": "^5.8.2"
  },
  "devDependencies": {
    "@cdktf/provider-aws": "^19.54.0",
    "@repo/eslint-config": "workspace:*",
    "@repo/typescript-config": "workspace:*",
    "@repo/vitest-config": "workspace:*",
    "@types/adm-zip": "^0.5.7",
    "@types/node": "^20.10.0",
    "@vitest/coverage-istanbul": "^3.0.9",
    "aws-cdk-lib": "^2.186.0",
    "cdktf": "^0.20.0",
    "constructs": "^10.4.2",
    "eslint": "^9.23.0",
    "mastra": "alpha",
    "terraconstructs": "^0.0.15",
    "tsx": "^4.19.3",
    "typescript": "5.8.2",
    "vitest": "^3.0.9"
  },
  "packageManager": "pnpm@9.15.1"
}
