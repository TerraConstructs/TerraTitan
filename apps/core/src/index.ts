import { select } from '@inquirer/prompts';
import { z } from 'zod';
import { argumentParser } from 'zodcli';
import { commandLineArgsSchema, awsCdkModules, listWorkflowRuns } from './cli-util.js';
import { runConversionWorkflow } from './wf-conversion.js';
import { runCdktfRefWf } from './wf-cdktf-ref.js';
import { runEnsureUpstreamWf } from './wf-ensure-upstream.js';
import { runBatchSourceConvertWf } from './wf-batch-source-convert.js';
import { runBatchTestConvertWf } from './wf-batch-test-convert.js';

import dotenv from 'dotenv';
dotenv.config();

const workflowTypeSchema = z.enum([
  /**
   * Current conversion workflow
   */
  'convert',
  /**
   * Demo Batch Source Code conversion workflow
   */
  'batchConvert',
  /**
   * Demo Batch Unit Tests conversion workflow
   */
  'batchTestConvert',
  /**
   * Demo CDKTF Reference retrieval workflow
   */
  'cdktfRef',
  /**
   * Demo ensure upstream workflow
   */
  'ensureUpstream',
  // 'convert-dynamic',
  // 'convert-vNext',
]);
async function main() {
  console.log('>>>TerraTitan CLI - PoC<<<');
  try {
    const args = argumentParser({ options: commandLineArgsSchema }).parse(process.argv.slice(2));
    if (args?.list) {
      return await listWorkflowRuns();
    }

    if (args?.module && !awsCdkModules.includes(args.module)) {
      throw new Error(`Invalid Module '${args.module}' not found in local aws-cdk-lib.`);
    }
    let wfToRun: z.infer<typeof workflowTypeSchema>;
    if (args.workflow) {
      const parsedWorkflow = workflowTypeSchema.safeParse(args.workflow);
      if (!parsedWorkflow.success) {
        throw new Error(`Invalid workflow: ${args.workflow}`);
      }
      wfToRun = parsedWorkflow.data;
    } else {
      // No workflow provided in args, prompt user to select wf
      wfToRun = await select<z.infer<typeof workflowTypeSchema>>({
        message: 'Select which workflow to run:',
        choices: [
          { value: 'convert', name: 'Conversion Workflow' },
          { value: 'batchConvert', name: 'Demo Batch Source Code Conversion Workflow' },
          { value: 'batchTestConvert', name: 'Demo Batch Unit Tests Conversion Workflow' },
          { value: 'cdktfRef', name: 'Demo aws-elasticloadbalancingv2.CfnTargetGroup CDKTF Reference retrieval' },
          { value: 'ensureUpstream', name: 'Demo ensure upstream for AWS CDK Module' },
          // { value: 'convert-dynamic', name: 'Conversion Wf (dynamic)' },
          // { value: 'convert-vNext', name: 'Conversion Wf vNext' },
        ],
      });
    }
    switch (wfToRun) {
      case 'convert':
        await runConversionWorkflow(args);
        break;
      case 'batchConvert':
        await runBatchSourceConvertWf();
        break;
      case 'batchTestConvert':
        await runBatchTestConvertWf();
        break;
      case 'cdktfRef':
        await runCdktfRefWf();
        break;
      case 'ensureUpstream':
        await runEnsureUpstreamWf(args);
        break;
      // case 'convert-dynamic':
      //   await runDynamicWf(args);
      //   break;
      // case 'convert-vNext':
      //   await runvNextWf(args);
      //   break;
      default:
        console.log('Invalid workflow selected.');
        break;
    }
  } catch (error) {
    if (error instanceof Error && error.name === 'ExitPromptError') {
      console.log('ðŸ‘‹ until next time!');
    } else {
      throw error;
    }
  }
}

main().catch(console.error);
