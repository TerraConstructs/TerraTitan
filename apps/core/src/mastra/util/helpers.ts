import path from 'path';
import { fileURLToPath } from 'url';
import { dirname } from 'path';
import { execSync } from 'child_process';
import * as ts from 'typescript';

// required for Hono / Dev Server ...
// __dirname is not defined in ES module scope
const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

// Get the git root directory
export const gitRoot = execSync('git rev-parse --show-toplevel', {
  encoding: 'utf8',
  cwd: __dirname,
}).trim();

/**
 * Get the base name of a typescript declaration file
 */
export function dtsBaseName(file: string): string {
  return path.basename(file, '.d.ts');
}

/**
 * Get the base name of a CDKTF ref file:
 *
 * for example:
 * "data/constructs/index.d.ts" => "constructs"
 */
export function cdktfBaseName(file: string): string {
  return path.basename(path.dirname(file));
}
/**
 * Simulate retrieval/filtering for prompting with relevant information
 *
 * Beware, AI Slop from this point forward...
 */

interface GeneratedImport {
  moduleSpecifier: string;
  namedSymbols: Set<string>;
  namespaceImport?: string;
}

/**
 * Find the generated imports in the source text
 */
export function findGeneratedImports(sourceText: string) {
  const sourceFile = ts.createSourceFile('temp.ts', sourceText, ts.ScriptTarget.Latest, true);
  const generatedImports: Map<string, GeneratedImport> = new Map();

  // First pass: collect import declarations from generated modules.
  ts.forEachChild(sourceFile, node => {
    if (ts.isImportDeclaration(node)) {
      // Remove quotes from module specifier.
      const moduleSpecifier = node.moduleSpecifier.getText(sourceFile).replace(/['"]/g, '');
      if (moduleSpecifier.includes('.generated')) {
        const genImport: GeneratedImport = {
          moduleSpecifier,
          namedSymbols: new Set(),
        };

        if (node.importClause) {
          const { namedBindings } = node.importClause;
          if (namedBindings) {
            if (ts.isNamedImports(namedBindings)) {
              // Handle named imports, e.g., import { CfnConnection } from './events.generated';
              for (const element of namedBindings.elements) {
                genImport.namedSymbols.add(element.name.getText(sourceFile));
              }
            } else if (ts.isNamespaceImport(namedBindings)) {
              // Handle namespace imports, e.g., import * as ssm from './ssm.generated';
              genImport.namespaceImport = namedBindings.name.getText(sourceFile);
            }
          }
        }
        generatedImports.set(moduleSpecifier, genImport);
      }
    }
  });

  // Second pass: traverse the AST to find property accesses on namespace imports.
  function findPropertyAccess(node: ts.Node) {
    if (ts.isPropertyAccessExpression(node)) {
      const expr = node.expression;
      if (ts.isIdentifier(expr)) {
        // For each generated import, check if the identifier matches the namespace alias.
        for (const genImport of generatedImports.values()) {
          if (genImport.namespaceImport === expr.getText(sourceFile)) {
            // Add the property name (e.g., CfnParameter) to the set of named symbols.
            genImport.namedSymbols.add(node.name.getText(sourceFile));
          }
        }
      }
    }
    ts.forEachChild(node, findPropertyAccess);
  }
  ts.forEachChild(sourceFile, findPropertyAccess);

  // Convert our Map into a plain object for easier consumption.
  const result: Record<string, string[]> = {};
  generatedImports.forEach((genImport, moduleSpecifier) => {
    result[moduleSpecifier] = Array.from(genImport.namedSymbols);
  });
  return result;
}

/**
 * Returns true if a declarationâ€™s name is one of the target symbols or related (e.g. "CfnConnectionProps")
 */
function isNameRelevant(name: string, filterSymbols: string[]): boolean {
  if (filterSymbols.includes(name)) {
    return true;
  }
  // Check for Props suffix and nested types
  for (const sym of filterSymbols) {
    if (name === `${sym}Props` || name.includes(`${sym}.`)) {
      return true;
    }
  }
  return false;
}

/**
 * Filters the given declaration source text so that only declarations related to
 * the provided symbols (classes, interfaces, functions, namespaces) are kept.
 *
 * @param sourceText The full declarations (as in a .d.ts file)
 * @param filterSymbols The list of symbol names (e.g. ["CfnConnection", "CfnRule"])
 */
export function filterGeneratedModule(sourceText: string, filterSymbols: string[]): string {
  const sourceFile = ts.createSourceFile('module.d.ts', sourceText, ts.ScriptTarget.Latest, true);

  // Transformer that recursively visits nodes and only keeps the ones we want.
  // The additional parameter 'inRelevantContext' tells us whether we're already inside
  // a relevant declaration (so we keep everything below it).
  const transformer = <T extends ts.Node>(context: ts.TransformationContext): ts.Transformer<T> => {
    function visit(node: ts.Node, inRelevantContext = false): ts.Node {
      // For container nodes, filter their statements.
      if (ts.isSourceFile(node) || ts.isModuleBlock(node)) {
        const filteredStatements = node.statements
          .map(st => visit(st, inRelevantContext))
          .filter((st): st is ts.Statement => st !== undefined);
        if (ts.isSourceFile(node)) {
          return ts.factory.updateSourceFile(node, filteredStatements);
        } else {
          return ts.factory.updateModuleBlock(node, filteredStatements);
        }
      }

      // Check for declarations: classes, interfaces, functions, and namespace/module declarations.
      if (
        ts.isClassDeclaration(node) ||
        ts.isInterfaceDeclaration(node) ||
        ts.isFunctionDeclaration(node) ||
        ts.isModuleDeclaration(node)
      ) {
        // If we're already inside a relevant declaration, keep everything.
        if (inRelevantContext) {
          return ts.visitEachChild(node, child => visit(child, true), context);
        }
        // Otherwise, if the declaration has a name and it is relevant...
        if (node.name && isNameRelevant(node.name.text, filterSymbols)) {
          // Mark children as relevant.
          return ts.visitEachChild(node, child => visit(child, true), context);
        }
        // Otherwise, check if any descendant declaration is relevant.
        let foundRelevant = false;
        function checkChild(n: ts.Node) {
          if (
            (ts.isClassDeclaration(n) ||
              ts.isInterfaceDeclaration(n) ||
              ts.isFunctionDeclaration(n) ||
              ts.isModuleDeclaration(n)) &&
            n.name &&
            isNameRelevant(n.name.text, filterSymbols)
          ) {
            foundRelevant = true;
          }
          ts.forEachChild(n, checkChild);
        }
        ts.forEachChild(node, checkChild);
        if (foundRelevant) {
          return ts.visitEachChild(node, child => visit(child, true), context);
        }
        // If neither the node nor any of its children are relevant, remove it.
        return ts.factory.createEmptyStatement();
      }
      // For other nodes, just continue recursing, preserving context.
      return ts.visitEachChild(node, child => visit(child, inRelevantContext), context);
    }
    return (node: T) => ts.visitNode(node, n => visit(n)) as T;
  };

  const result = ts.transform(sourceFile, [transformer]);
  const transformedSourceFile = result.transformed[0] as ts.SourceFile;
  const printer = ts.createPrinter();
  const output = printer.printFile(transformedSourceFile);
  result.dispose();

  // If the output only contains semicolons or whitespace, return empty string
  return output.replace(/^[\s;]*$/, '');
}

/**
 * Process input to extract and filter relevant declarations from inputRef
 */
export function filterInputRefFile(inputSource: string, inputRefSource: string): string {
  const importsByModule = findGeneratedImports(inputSource);

  // TODO: Validate the inputRefSource has all the necessary declarations

  // Collect all symbols from all generated imports
  const symbolsToFilter = Object.values(importsByModule).flatMap(symbols => symbols);

  // Filter the input reference file to only include relevant declarations
  const filteredDeclarations = filterGeneratedModule(inputRefSource, symbolsToFilter);

  return filteredDeclarations;
}
