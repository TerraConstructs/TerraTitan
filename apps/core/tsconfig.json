{
  "extends": ["@repo/typescript-config/node.json"],
  "compilerOptions": {
    "outDir": "./dist",
  },
  "include": ["src/**/*"],
  "exclude": ["node_modules", "**/*.test.ts", "**/*.ts.snap", "dist", "data"]
}
