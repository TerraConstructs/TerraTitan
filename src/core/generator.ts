import { Anthropic } from "@anthropic-ai/sdk";
import { AwsCdkConstruct, FileType } from "../types";
import { generateWithClaude } from "../anthropic/client";

/**
 * Generate Terraform CDK code based on AWS CDK construct
 * @param parsedConstruct Parsed AWS CDK construct
 * @param similarPatterns Similar patterns from vector DB
 * @param anthropicClient Anthropic client instance
 * @returns Generated Terraform CDK code
 */
export async function generateTerraformCdk(
  parsedConstruct: AwsCdkConstruct,
  similarPatterns: any[] = [],
  anthropicClient: Anthropic
): Promise<string> {
  if (parsedConstruct.fileType === FileType.TEST) {
    return generateTestCode(parsedConstruct, similarPatterns, anthropicClient);
  } else {
    return generateSourceCode(
      parsedConstruct,
      similarPatterns,
      anthropicClient
    );
  }
}

async function generateSourceCode(
  parsedConstruct: AwsCdkConstruct,
  similarPatterns: any[] = [],
  anthropicClient: Anthropic
): Promise<string> {
  const prompt = `Convert this AWS CDK construct to Terraform CDK:

Source code:
\`\`\`typescript
${parsedConstruct.source}
\`\`\`

${
  similarPatterns.length > 0
    ? "Here are similar conversion patterns for reference:"
    : ""
}
${similarPatterns
  .map(
    (pattern) => `\`\`\`typescript\n${JSON.stringify(pattern, null, 2)}\n\`\`\``
  )
  .join("\n\n")}

Please convert the above AWS CDK construct to Terraform CDK (CDKTF) code.
Keep the same functionality but use Terraform CDK constructs and patterns.`;

  return await generateWithClaude(prompt, {
    system:
      "You are an expert in both AWS CDK and Terraform CDK. Convert the provided AWS CDK construct to Terraform CDK, preserving all functionality.",
  });
}

async function generateTestCode(
  parsedConstruct: AwsCdkConstruct,
  similarPatterns: any[] = [],
  anthropicClient: Anthropic
): Promise<string> {
  const prompt = `Convert this AWS CDK Jest test file to Terraform CDK:

Test code:
\`\`\`typescript
${parsedConstruct.source}
\`\`\`

${
  similarPatterns.length > 0
    ? "Here are similar conversion patterns for reference:"
    : ""
}
${similarPatterns
  .map(
    (pattern) => `\`\`\`typescript\n${JSON.stringify(pattern, null, 2)}\n\`\`\``
  )
  .join("\n\n")}

Please convert the above AWS CDK Jest tests to Terraform CDK (CDKTF) tests.
Keep the same test coverage and assertions but updated for Terraform CDK patterns.`;

  return await generateWithClaude(prompt, {
    system:
      "You are an expert in both AWS CDK and Terraform CDK testing. Convert the provided AWS CDK Jest tests to Terraform CDK tests, preserving all test coverage and assertions.",
  });
}
