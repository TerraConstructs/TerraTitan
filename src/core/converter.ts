import { parseAwsCdkConstruct } from "./parser";
import { generateTerraformCdk } from "./generator";
import { validateConversion } from "./validator";
import { getVectorDbClient } from "../vectordb/client";
import { getAnthropicClient } from "../anthropic/client";
import {
  ConversionResult,
  AwsCdkConstruct,
  ConversionOptions,
  FileType,
} from "../types";
import { detectFileType } from "../utils/fileUtils";

/**
 * Convert an AWS CDK construct to Terraform CDK
 *
 * @param sourceFilePath Path to the AWS CDK construct file
 * @param options Conversion options
 * @returns The conversion result with Terraform CDK code
 */
export async function convertConstruct(
  sourceFilePath: string,
  options: ConversionOptions = {}
): Promise<ConversionResult> {
  // Parse the AWS CDK construct
  const parsedConstruct = await parseAwsCdkConstruct(sourceFilePath);

  // Override file type if specified in options
  if (options.fileType) {
    parsedConstruct.fileType = options.fileType;
  }

  // Initialize Anthropic client
  const anthropicClient = getAnthropicClient();

  // Initialize Vector DB client
  const vectorDbClient = getVectorDbClient();

  // Find similar patterns in vector database
  const similarPatterns = await vectorDbClient.findSimilarPatterns(
    parsedConstruct
  );

  // Generate Terraform CDK code using Anthropic
  const generatedCode = await generateTerraformCdk(
    parsedConstruct,
    similarPatterns,
    anthropicClient
  );

  // Validate the generated code
  const validationResult = await validateConversion(
    parsedConstruct,
    generatedCode
  );

  // Store the new pattern if validation passes and not a test file
  if (
    validationResult.success &&
    options.storePattern !== false &&
    parsedConstruct.fileType !== FileType.TEST
  ) {
    await vectorDbClient.storePattern(parsedConstruct, generatedCode);
  }

  return {
    originalConstruct: parsedConstruct,
    terraformCdkCode: generatedCode,
    validationResult,
    similarPatterns,
  };
}
