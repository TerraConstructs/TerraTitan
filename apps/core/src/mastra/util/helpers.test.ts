import { expect, test, describe } from 'vitest';
import { filterGeneratedModule, findGeneratedImports, filterInputRefFile, kebabToTitleCase } from './helpers.js';

describe('kebabToTitleCase', () => {
  test('should convert kebab-case to TitleCase', () => {
    const input = 'sns-topic';
    const expectedOutput = 'SnsTopic';
    const result = kebabToTitleCase(input);
    expect(result).toEqual(expectedOutput);
  });
  test('should handle empty string', () => {
    const input = '';
    const expectedOutput = '';
    const result = kebabToTitleCase(input);
    expect(result).toEqual(expectedOutput);
  });
  test('should handle single word', () => {
    const input = 'example';
    const expectedOutput = 'Example';
    const result = kebabToTitleCase(input);
    expect(result).toEqual(expectedOutput);
  });
});

describe('filterInputRefFile', () => {
  test('should process input file and filter declarations', async () => {
    const inputSource = `
      import { CfnConnection } from './events.generated';
      import * as ssm from './ssm.generated';

      const connection = new CfnConnection();
      const resource = new ssm.CfnParameter(this, 'Resource', {});
    `;

    const inputRefSource = `
      export declare class CfnConnection {
        public readonly x: string;
      }
      export interface CfnConnectionProps {
        readonly prop1: string;
      }
      export declare class OtherClass {
        public readonly y: string;
      }
    `;
    const result = filterInputRefFile(inputSource, inputRefSource);

    expect(result).toContain('export declare class CfnConnection');
    expect(result).toContain('export interface CfnConnectionProps');
    expect(result).not.toContain('export declare class OtherClass');
  });
});

describe('findGeneratedImports', () => {
  test('should find generated imports', () => {
    const sourceText = `
      import { CfnConnection } from './events.generated';
      import * as ssm from './ssm.generated';

      const connection = new CfnConnection();
      const resource = new ssm.CfnParameter(this, 'Resource', {
        allowedPattern: 'abc',
        description: 'A parameter',
      });
    `;

    const result = findGeneratedImports(sourceText);

    expect(result).toEqual({
      './events.generated': ['CfnConnection'],
      './ssm.generated': ['CfnParameter'],
    });
  });

  test('should handle multiple named imports', () => {
    const sourceText = `
      import { CfnConnection, CfnRule } from './events.generated';
    `;

    const result = findGeneratedImports(sourceText);

    expect(result).toEqual({
      './events.generated': ['CfnConnection', 'CfnRule'],
    });
  });

  test('should handle namespace imports with property access', () => {
    const sourceText = `
      import * as cdk from './cdk.generated';
      cdk.Stack.create();
      cdk.App.build();
    `;

    const result = findGeneratedImports(sourceText);

    expect(result).toEqual({
      './cdk.generated': ['Stack', 'App'],
    });
  });
});

describe('filterGeneratedModule', () => {
  test('should keep declarations for specified symbols and their related types', () => {
    const generatedDeclaration = `
import * as cdk from "../../core";
import * as constructs from "constructs";

export declare class CfnApiDestination extends cdk.CfnResource implements cdk.IInspectable {
  public readonly x: string;
}

export interface CfnApiDestinationProps {
  readonly prop1: string;
}

export declare class CfnConnection extends cdk.CfnResource implements cdk.IInspectable {
  public readonly y: string;
}

export declare namespace CfnConnection {
  interface AuthParametersProperty {
    readonly apiKeyAuthParameters?: ApiKeyAuthParametersProperty;
  }
  interface ApiKeyAuthParametersProperty {
    readonly apiKeyName: string;
    readonly apiKeyValue: string;
  }
}

export interface CfnConnectionProps {
  readonly prop2: string;
}`;

    const result = filterGeneratedModule(generatedDeclaration, ['CfnConnection']);

    // Verify CfnConnection and related declarations are kept
    expect(result).toMatchSnapshot();
    expect(result).toContain('export declare class CfnConnection');
    expect(result).toContain('export interface CfnConnectionProps');
    expect(result).toContain('export declare namespace CfnConnection');
    expect(result).toContain('interface AuthParametersProperty');
    expect(result).toContain('interface ApiKeyAuthParametersProperty');

    // Verify unrelated declarations are removed
    expect(result).not.toContain('CfnApiDestination');
    expect(result).not.toContain('CfnApiDestinationProps');
  });

  test('should handle multiple symbols', () => {
    const generatedDeclaration = `
export declare class CfnRule extends Resource {
  public readonly x: string;
}

export interface CfnRuleProps {
  readonly prop1: string;
}

export declare class CfnConnection extends Resource {
  public readonly y: string;
}

export declare class OtherClass extends Resource {
  public readonly z: string;
}`;

    const result = filterGeneratedModule(generatedDeclaration, ['CfnConnection', 'CfnRule']);

    expect(result).toContain('export declare class CfnConnection');
    expect(result).toContain('export declare class CfnRule');
    expect(result).toContain('export interface CfnRuleProps');
    expect(result).not.toContain('export declare class OtherClass');
  });

  test('should handle empty input', () => {
    const result = filterGeneratedModule('', ['CfnConnection']);
    expect(result).toBe('');
  });

  test('should handle cases where no symbols match', () => {
    const generatedDeclaration = `
export declare class SomeClass {
  public readonly x: string;
}`;

    const result = filterGeneratedModule(generatedDeclaration, ['NonExistentSymbol']);
    expect(result.trim()).toBe('');
  });
});
