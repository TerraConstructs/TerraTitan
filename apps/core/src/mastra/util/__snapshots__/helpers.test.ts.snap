// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`filterGeneratedModule > should keep declarations for specified symbols and their related types 1`] = `
"import * as cdk from "../../core";
import * as constructs from "constructs";
;
;
export declare class CfnConnection extends cdk.CfnResource implements cdk.IInspectable {
    public readonly y: string;
}
export declare namespace CfnConnection {
    interface AuthParametersProperty {
        readonly apiKeyAuthParameters?: ApiKeyAuthParametersProperty;
    }
    interface ApiKeyAuthParametersProperty {
        readonly apiKeyName: string;
        readonly apiKeyValue: string;
    }
}
export interface CfnConnectionProps {
    readonly prop2: string;
}
"
`;
